singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -248,19 +249,89 @@ class PTESolverBase {
   }
 
   PORTABLE_FORCEINLINE_FUNCTION
-  void SetVfracFromT(const Real T) {
-    Real vsum = 0.0;
-    // set volume fractions
+  void NormalizeVfrac() {
+    Real vfrac_sum = 0;
+    for (std::size_t m = 0; m < nmat; ++m) {
+      vfrac_sum += vfrac[m];
+    }
     for (std::size_t m = 0; m < nmat; ++m) {
+      vfrac[m] *= robust::ratio(vfrac_total, vfrac_sum);
+    }
+  }
+
+  /* JMM: Find the volume fractions that:
+   * 1) Best respect initial guesses
+   * 2) Sum to 1
+   * 3) Are bounded from above such that no material is in tension (if
+   * possible)
+   *
+   * If these are not all possible to respect, error out and expect
+   * the GetTguess function to save us by picking a higher
+   * temperature.
+   *
+   * If this is possible to achieve, we can do so by using the "water
+   * filling" algorithm. See, e.g., Gallager, 1968.
+   */
+  PORTABLE_FORCEINLINE_FUNCTION
+  bool SetVfracFromT(const Real T) {
```

Comment:
The diff is kind of confusing here. I think just try to ignore the red. I completely rewrote this function.